lib/SILOptimizer/Utils/Local.cpp

```diff
@@ -1677,6 +1678,25 @@ optimizeBridgedSwiftToObjCCast(SILInstruction *Inst,
   SILType SubstFnTy = SILFnTy.substGenericArgs(M, SubMap);
   SILFunctionConventions substConv(SubstFnTy.castTo<SILFunctionType>(), M);
 
+  if (Dest) {
+    // check that we can go through with the optimization
+    auto ConvTy = substConv.getSILResultType().getObjectType();
+    auto DestTy = Dest->getType().getObjectType();
+    if ((ConvTy != DestTy) && !(DestTy.isExactSuperclassOf(ConvTy)) &&
```

Comment:
IMO, this complex condition is difficult to understand. Can you extract the whole check into a separate function, which lets you split the condition into multiple if-return statements?